Fix alias ABI and docs

cgmb opened the pull request:
This change fixes a number of problems relating to the rocSOLVER aliases.

These aliases are now marked as deprecated in the documentation, and the code has been annotated so GCC or Clang can issue deprecation warnings to users. In the process, I fixed a number of build warnings and errors in the documentation (and made a number of other minor documentation improvements).

The aliases are also now treated as regular API functions. I've added a very basic sample program to ensure that they are accessible from Fortran. It doesn't seem particularly important to ensure deprecated functions have accessible symbols, but I hope the infrastructure that was built for that will be useful for testing more important functionality in the future.

cgmb commented:
The CI errors appear to all be on the Ubuntu build, which is undergoing infrastructure upgrades. It doesn't look like Jenkins even tried to build the library, so I believe those are spurious failures.

saadrahim commented:
The CI error is not an infrastructure upgrade. Please try the CI build again and see if it works. It should pass on Ubuntu or please assist me in root causing the of the failure.

cgmb commented:
My mistake! The underlying problem was that the downloaded rocBLAS artifact was empty. I just made a trivial change to trigger a rebuild.

cgmb commented:
Correction: The artifact was corrupt, not empty. In any case, it appears to have been a transient failure.

saadrahim commented:
You do not need to make a change to trigger a rebuild. I will try to arrange a introduction session to go over commonly used CI features.

jzuniga-amd commented on rocsolver/clients/samples/example_deprecated_aliases.f90:
```diff
@@ -0,0 +1,202 @@
+
+    !!!!!!!!!!!!!!!!!!!!!!!
+    !    rocSOLVER types    !
+    !!!!!!!!!!!!!!!!!!!!!!!
```
Why show a code sample with features that are deprecated (the rocsolver aliases) ?

cgmb replied:
I was just making sure my changes to the ABI worked, but you're right it's kind of silly. I've replaced it with a very basic Fortran example that does a QR factorization.

jzuniga-amd commented on rocsolver/clients/samples/example_simple.cpp:
```diff
@@ -0,0 +1,44 @@
+  // here is where you would initialize M, N and lda with desired values
+
+  rocsolver_handle handle;
+  rocsolver_create_handle(&handle);
```
Similarly here, I think that if the deprecation is documented and you are adding warnings to these functions, we should be using rocblas types and auxiliaries everywhere. ??

cgmb replied:
This sample is identical to the samples in the docs, so I'll update those too. I'd partially updated them in #120, but wasn't sure about `rocsolver_create_handle` at the time.

cgmb replied:
I've updated this sample to avoid deprecated types, and added some actual data so the sample can be run without invoking undefined behaviour.

jzuniga-amd commented on rocsolver/docs/source/userguide_api.rst:
```diff
@@ -13,71 +13,89 @@ Types
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 .. doxygentypedef:: rocsolver_operation
+.. deprecated:: 3.5
+   Use :c:enum:`rocblas_operation`.
```
I am getting warning when building the docs:
/home/jzuniga-/rocSOLVER/rocsolver/docs/source/userguide_api.rst:40: WARNING: Unknown interpreted text role "c:enum".
/home/jzuniga-/rocSOLVER/rocsolver/docs/source/userguide_api.rst:46: WARNING: Unknown interpreted text role "c:enum".
/home/jzuniga-/rocSOLVER/rocsolver/docs/source/userguide_api.rst:52: WARNING: Unknown interpreted text role "c:enum".
/home/jzuniga-/rocSOLVER/rocsolver/docs/source/userguide_api.rst:58: WARNING: Unknown interpreted text role "c:enum".
/home/jzuniga-/rocSOLVER/rocsolver/docs/source/userguide_api.rst:64: WARNING: Unknown interpreted text role "c:enum".
/home/jzuniga-/rocSOLVER/rocsolver/docs/source/userguide_api.rst:70: WARNING: Unknown interpreted text role "c:enum".
/home/jzuniga-/rocSOLVER/rocsolver/docs/source/userguide_api.rst:76: WARNING: Unknown interpreted text role "c:enum".

It shows like this in the generated documents:
Deprecated since version 3.5: Use **:c:enum:`rocblas_status`**.

cgmb replied:
I think you may be missing the python3 `breathe` package from pip. It was already listed in `rocsolver/docs/source/requirements.txt`, but I'm not sure if there's somewhere else it needs to be specified.

I found it a little confusing trying to figure out what process was required build the docs, and I'm still pondering over what might be the best way to document that for posterity. IMO, a local build of the docs should be included in the CI build and maybe added as a flag to `install.sh`. Automation isn't necessarily the easiest documentation to read, but it's thorough and it tests itself.

I wrote [a Dockerfile](https://github.com/cgmb/rocSOLVER/blob/test-build-docs/docker/dockerfile-docs) based on my guesses at how we expect to use the readthedocs build environment. It was just a discussion piece, though, so I didn't include it in this PR.

cgmb replied:
I didn't change anything here, but it builds correctly for me in that docker container.

jzuniga-amd replied:
Yeah, maybe something is missing on my setup. The important thing is that ReadTheDocs can generate the documents correctly (which I expect to be the case). Once this is merged, we can verify https://rocsolver.readthedocs.io/en/latest/.

jzuniga-amd commented on rocsolver/docs/source/userguide_api.rst:
```diff
@@ -13,71 +13,89 @@ Types
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. doxygentypedef:: rocsolver_direction
+.. deprecated:: 3.5
+   Use :c:enum:`rocblas_direction`.
```
rocsolver/rocblas_direct is not defined in rocBLAS. It is an extra type only used in rocSOLVER. We should include its documentation from rocsolver/library/include/rocsolver-extra-types.h

cgmb replied:
I definitely was confused with these ones. I just deleted `rocsolver_direct`... though looking at your comment again, perhaps that was a mistake too. Hmm... We should briefly chat about what the right thing to do here is.

jzuniga-amd replied:
What you are doing now is correct!!

jzuniga-amd commented on rocsolver/docs/source/userguide_api.rst:
```diff
@@ -13,71 +13,89 @@ Types
+.. doxygentypedef:: rocsolver_storev
+.. deprecated:: 3.5
+   Use :c:enum:`rocblas_storev`.
 
```
Similarly for _storev. It is not defined/used in rocBLAS. It is defined in rocsolver/library/include/rocsolver-extra-types.h and its information should be included in the documentation.

jzuniga-amd commented on rocsolver/docs/source/userguide_install.rst:
```diff
@@ -10,26 +10,24 @@ Building and installation
-branch tip as well.)
+rocBLAS and rocSOLVER are both still under active development, and it is hard to define minimal
+compatibility versions. For now, a good rule of thumb is to always use rocSOLVER the matching
+rocBLAS version. For example, if you want to install rocSOLVER from ROCm 3.3 release, then be sure
```
"For now, a good rule of thumb is to always use rocSOLVER the matching rocBLAS version" ??
Something is missing? Maybe "... use rocSOLVER **with** the matching rocBLAS version" ??

cgmb replied:
Reworded.

tfalders commented on rocsolver/docs/source/userguide_install.rst:
```diff
@@ -76,7 +74,7 @@ in the output directory:
 and building process can also be found      
 at ``rocSOLVER/build`` and ``rocSOLVER/build/release`` directories.  
-It is assummed that all the library external
+It is assumed that all the library external
```
Suggested correction: ... all external library dependencies ...

cgmb replied:
I've reworded this.

cgmb commented on rocsolver/docs/source/userguide_install.rst:
```diff
@@ -10,26 +10,24 @@ Building and installation
-are building the rocSOLVER branch tip, then you will need to build and install rocBLAS
-branch tip as well.)
+rocBLAS and rocSOLVER are both still under active development, and it is hard to define minimal
+compatibility versions. For now, a good rule of thumb is to always use rocSOLVER the matching
```
Reworded.

cgmb commented on rocsolver/docs/source/userguide_intro.rst:
```diff
@@ -10,21 +10,21 @@ Introduction
 
-rocSolver Library is in early stages of active development. New features and functionality is being continuosly added. New 
-functionality is documented at each release of the ROCm platform. 
+rocSolver Library is in early stages of active development. New features and functionality is being
```
I've reworked the paragraph.

cgmb commented on rocsolver/clients/samples/example_simple.cpp:
```diff
@@ -0,0 +1,44 @@
+  // here is where you would initialize M, N and lda with desired values
+
+  rocsolver_handle handle;
+  rocsolver_create_handle(&handle);
```
I've replaced them.

cgmb commented:
I *think* I have addressed all the concerns raised. I'm still a little unsure if I handled `rocblas_direct`/`rocsolver_direction` and `rocsblas_storev`/`rocsolver_storev` correctly.

The docs can be built in a docker container defined like so:
```
FROM readthedocs/build:latest

USER root:root
RUN apt-get install -qq doxygen
RUN pip3 install breathe sphinx sphinx_rtd_theme
COPY --chown=docs:docs . /home/docs

USER docs:docs
WORKDIR /home/docs/rocsolver/docs
RUN sh run_doc.sh
```

Usually, I generate the docs using a script:
```
docker build --tag docs:latest .
container_id="$(docker create docs:latest)"
docker cp "$container_id:/home/docs/rocsolver/docs" docs
```

jzuniga-amd commented on rocsolver/clients/samples/example_basic.cpp:
```diff
@@ -0,0 +1,91 @@
+
+  // let's print the input matrix, just to see it
+  printf("A = [\n");
+  for (size_t i = 0; i < lda; ++i) {
```
I would use i < m. 
This way is clear that you are travelling the entries of the matrix mapped by hA. Entries between m and lda are not touched by the algorithms (they contain garbage in general) and should not be accessed this way. In this case, if lda > m you will be printing stuff that does not belong to the matrix mapped by hA.

cgmb replied:
I wasn't sure about that. Thanks! Fixed.

jzuniga-amd commented on rocsolver/clients/samples/example_basic.cpp:
```diff
@@ -0,0 +1,91 @@
+  // the results are now in hA and hIpiv
+  // we can print some of the results if we want to see them
+  printf("R = [\n");
+  for (size_t i = 0; i < lda; ++i) {
```
same here...

jzuniga-amd commented on rocsolver/docs/source/userguide_examples.rst:
```diff
@@ -9,73 +9,119 @@ Using rocSOLVER
+      hA.resize(size_t(lda) * N);
+      for (size_t i = 0, j = 0; j < N; ++j) {
+        for (size_t k = 0; k < lda; ++k, ++i) {
+          hA[i] = A[k][j]; // convert A (row-major) for hA (column-major)
```
same comments as above

jzuniga-amd commented on rocsolver/docs/source/userguide_examples.rst:
```diff
@@ -9,73 +9,119 @@ Using rocSOLVER
+      // let's print the input matrix, just to see it
+      printf("A = [\n");
+      for (size_t i = 0; i < lda; ++i) {
+        printf("  ");
```
same comments as above

jzuniga-amd commented on rocsolver/docs/source/userguide_examples.rst:
```diff
@@ -9,73 +9,119 @@ Using rocSOLVER
+      // we can print some of the results if we want to see them
+      printf("R = [\n");
+      for (size_t i = 0; i < lda; ++i) {
+        printf("  ");
```
same comments as above

jzuniga-amd commented:
I added the noOptimizations label. This tells CI to build this PR with no optimization (./install.sh -n) which speeds up the process. You can use this flag in all PRs that do not touch the small and mid-size optimization code.

jzuniga-amd commented:
You should rebase your branch to include the latest commits to develop (in particular Daine's commit) if you want CI tests to pass.

cgmb commented on rocsolver/clients/samples/example_basic.cpp:
```diff
@@ -0,0 +1,91 @@
+  for (size_t i = 0, j = 0; j < N; ++j) {
+    for (size_t k = 0; k < lda; ++k, ++i) {
+      hA[i] = A[k][j]; // convert A (row-major) for hA (column-major)
+    }
```
I generally just think of all C array types as just being blobs of data with different accessors, as you can always access a 2D array like a decayed 1D array just by grabbing a pointer to the first element. (And if C++ stored things in column-major order, that's how I would have done the copy.)

However, I'm sure that you think much more similarly to the folks who will be using this code. Changed.

jzuniga-amd commented:
When ready, remember to use "squash and merge".
Thanks Cory!!!
